live-tests/macros/src/lib.rs

```diff
@@ -0,0 +1,80 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Macro to wrap a live test function that automatically creates and cleans up
+//! the `LiveTestContext`
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, ItemFn};
+
+/// Attribute for wrapping a test function to handle automatically creating and
+/// destroying a LiveTestContext. If the wrapped test fails, the context will
+/// intentionally not be cleaned up to support debugging.
+///
+/// Example usage:
+///
+/// ```ignore
+/// #[live_test]
+/// async fn test_my_test_case(lc: &mut LiveTestContext) {
+///   assert!(true);
+/// }
+/// ```
+///
+/// We use this instead of implementing Drop on LiveTestContext because we want
+/// the teardown to only happen when the test doesn't fail (which causes a panic
+/// and unwind).
```

Comment:
I went down this path in a follow-on branch:
https://github.com/oxidecomputer/omicron/pull/new/dap/experiments/live-tests-no-macro

A few notes:

- I wanted to avoid type parameters in `LiveTestContext::run_test` because I didn't want it monomorphized for every test that we add.  So I had `run_test` accept a `&dyn Fn` instead of `F: Fn` and that closure returns a `BoxFuture`.
- rustfmt seems to have lost its mind on the code and it looks awful.  I have spent no time trying to figure out why.
- It doesn't quite compile due to a lifetime issue.  I gave up (not at all insurmountable, I was just spending more time on it than I had hoped to for just an experiment).

There are a couple of problems:

- the rustfmt problem, but I'll assume that is surmountable
- it's much easier to copy/paste the code and get the wrong `test_name` in the argument
- `[tokio::test]` and so `[live_test]` allow your test function to return nothing at all or a Result type.  I had this version return a `Result`.  It would be tricky to support both here.  (But most of our tests probably just return `()` and we could probably just have it support that.)

In summary, I don't like it better but I'm still open to it if people really do!

Reply:
Oh thanks! Looking at https://github.com/oxidecomputer/omicron/commit/3f43f781e154d8c888b146947a83b91370adff33:

* The rustfmt issue is because of the additional nesting involved afaict.
* I think monomorphizing that is okay? The function looks pretty small to me, and the dependent functions aren't generic.
3. Passing in borrowed data to closures is difficult. [async closures](https://github.com/rust-lang/rust/issues/62290) solve this but aren't stable yet. I ran into this with the update engine, and that has to pass an owned context in for that reason.